### PR TITLE
[JENKINS-65288] Bring back support for stapler.jelly.trace

### PR DESCRIPTION
**Pull request by Wadeck**
See [JENKINS-65288](https://issues.jenkins-ci.org/browse/JENKINS-65288).
Also a comment in [JENKINS-28561](https://issues.jenkins.io/browse/JENKINS-28561).

I tried to re-use `-Dstapler.jelly.trace=true` (coming from [Stapler](https://github.com/stapler/stapler/blob/c8527cc9387c95b5eaa22a4207e15df18c5280b8/jelly/src/main/java/org/kohsuke/stapler/jelly/JellyFacet.java#L127)) and it fails. This system property helps undertanding where the Jelly fragments are coming from by adding comments around the fragments.

There are two reasons for the failure.

**1) ❌ firstChild 🚸**
The firstChild JavaScript method is not expected to return a CommentNode in our JS. `firstChild` can return Element, TextNode or CommentNode, while `firstElementChild` can only return Element.
More details in https://developer.mozilla.org/en-US/docs/Web/API/Node/firstChild, especially the paragraph comparing it with firstElementChild.

From the core code I checked, the Element seems to be the only one really desired, but I can be wrong. 
@timja @fqueiruga please review this PR for this :) 

**2) ❌ Flush 🚽**
⚠️ More complicated to explain, the second point is due to the correction we applied to SECURITY-534, i.e. the dispatch allowance introduced to prevent Jelly fragments to be rendered as a standalone page.
The comment from the system property are [directly written to the output](https://github.com/stapler/stapler/blob/0a23a9000de1dafb6a16fa883504697443cc05e3/jelly/src/main/java/org/kohsuke/stapler/jelly/CallTagLibScript.java#L112-L115). By itself it has no impact, but the presence of content inside the output means that the next call to `flush`, a real flush is executed. This is due to the implementation of the StreamEncoder#implFlushBuffer, if there is something in `bb`, the flush is done.
Why flush is important? It's because in the correction, we waited for the first flush before checking if the dispatch is allowed or not. This is either allowed programmatically or by checking the status/contentType of the response. This logic is done in [DefaultScriptInvoker.java#L117-L120](https://github.com/stapler/stapler/blob/c8527cc9387c95b5eaa22a4207e15df18c5280b8/jelly/src/main/java/org/kohsuke/stapler/jelly/DefaultScriptInvoker.java#L117-L120).

The `ajax.jelly` modifications are there to ensure no tags are forcing a flush before the contentType is set. Lots of Jelly/Stapler tags (e.g. st:setHeader or j:choose) are [flushing by default](https://github.com/jenkinsci/jelly/blob/cb7966734f339619e5ee8e57b08009acbf5e1d10/src/java/org/apache/commons/jelly/impl/TagScript.java#L267), but this is not the case for [j:set](https://github.com/jenkinsci/jelly/blob/cb7966734f339619e5ee8e57b08009acbf5e1d10/src/java/org/apache/commons/jelly/tags/core/CoreTagLibrary.java#L109) and [j:if](https://github.com/jenkinsci/jelly/blob/cb7966734f339619e5ee8e57b08009acbf5e1d10/src/java/org/apache/commons/jelly/tags/core/CoreTagLibrary.java#L94). Due to this change, the first flush call is done after the contentType is set and thus, the dispatch is now allowed.

For this part, please @jvz review the PR.

### Testing notes 🧪 

**A) Ensure the correction is working**
Checkout this PR, start the WAR with `-Dstapler.jelly.trace=true`, do in parallel the same with uncorrect Jenkins WAR. Test anything you can imagine, the behavior should be working on the WAR from this PR and potentially not in the regular one.
Example of broken features:
- ajaxExecutors and like ajax requests are broken
- Global Security configuration, Job configuration and Tool configuration submissions are broken

**B) Ensure there is no regression**
Checkout this PR, start the WAR regularly (without the system property), do the same in parallel with a regular Jenkins WAR. All the features working in the regular build should also work on this PR's WAR.

I didn't find anything broken during my local testing.

### Proposed changelog entries

* Developer: `stapler.jelly.trace` support is back

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

✔️ Already mentioned in their related parts.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by jvz — COMMENTED** [review]
The fix sounds legit, good find! I'll be back later with a more thorough review.

**Review by jvz — APPROVED** [review]
Tested with master and this PR. This does indeed fix the various broken JavaScript when enabling request trace logging. Verified the ajax executors URL which would otherwise give log messages like:

`2021-04-05 16:11:27.625+0000 [id=83]	WARNING	j.s.s.StaplerFilteredActionListener#onDispatchTrigger: New Stapler dispatch rules result in the URL "/ajaxExecutors" no longer being allowed. If you consider it safe to use, add the following to the whitelist: "hudson.model.AllView ajaxExecutors". Learn more: https://jenkins.io/redirect/stapler-facet-restrictions`

And the fun JSON parse error on the security configuration page was indeed fixed here. Other pages generally seem to be working properly with tracing.

**Review by timja — APPROVED** [review]
Can’t see any issues, have not run the code

File: test/src/test/java/hudson/security/SecurityRealmTest.java [hunk 1]
```diff
@@ -25,9 +25,12 @@
 
 import com.gargoylesoftware.htmlunit.CookieManager;
 import com.gargoylesoftware.htmlunit.WebResponse;
+import com.gargoylesoftware.htmlunit.html.HtmlElementUtil;
+import com.gargoylesoftware.htmlunit.html.HtmlForm;
 import com.gargoylesoftware.htmlunit.html.HtmlPage;
 import com.gargoylesoftware.htmlunit.util.Cookie;
 import hudson.model.Descriptor;
+import hudson.model.FreeStyleProject;
```
Comment (daniel-beck) [inline test/src/test/java/hudson/security/SecurityRealmTest.java:33]: Unused?

File: test/src/test/java/hudson/security/SecurityRealmTest.java [hunk 2]
```diff
@@ -25,9 +25,12 @@
 
 import com.gargoylesoftware.htmlunit.CookieManager;
 import com.gargoylesoftware.htmlunit.WebResponse;
+import com.gargoylesoftware.htmlunit.html.HtmlElementUtil;
```
Comment (daniel-beck) [inline test/src/test/java/hudson/security/SecurityRealmTest.java:28]: Unused?

File: core/src/main/resources/lib/layout/ajax.jelly [hunk 3]
```diff
@@ -42,18 +42,22 @@ THE SOFTWARE.
...
   </st:documentation>
-  <st:setHeader name="Cache-Control" value="no-cache,no-store,must-revalidate" />
-  <j:choose>
-    <j:when test="${rootURL!=null}">
+  <!-- 
+    j:if tag is implemented without using TagScript like j:choose and thus, does not trigger the flush of the output
+    this has a consequence when using -Dstapler.jelly.trace=true, which adds comments
+  -->
+  <j:set var="standalone" value="${rootURL==null}"/>
+  <j:if test="${!standalone}">
```
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/ajax.jelly:50]: Probably safe to no longer `Cache-Control` here, since this isn't a standalone element.

**Review by daniel-beck — APPROVED** [review]
Test look reasonable. Not tested manually.

**Review by oleg-nenashev — APPROVED** [review]
Looks good. Do we also want to backport it into the 2.277.x LTS line? Arguably, this is an old regression which may impact developers and advanced users who troubleshoot their instances.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by daniel-beck** [comment]
> Looks good. Do we also want to backport it into the 2.277.x LTS line?

No.

**Comment by Wadeck** [comment]
Thanks for the review, just removed 2 unused imports. 🚀
